### DISCUSSION TITLE
-ActivatedAction not working in new powershell instance

### DISCUSSION THREAD

**Discussion by iTob191** [Windos/BurntToast#145] [Q&A] [answered]
Hey

I finally got around to updating a script to use `-ActivatedAction` (thanks for adding that!). Minimal example:
```ps1
# script.ps1
New-BurntToastNotification `
	-Text "text" `
	-ActivatedAction {
		start "cmd" "/c echo Hello && pause"
	}
```
This works fine, when executing the script directly (`.\script.ps1`). When started in a new powershell instance (`pwsh .\script.ps1`) however, the notification gets shown as expected, but clicking on it does not do anything. I suppose this could be solved by using a protocol handler, but is there an easier solution that does not involve adding registry keys and so on?

Thanks in advance!

PowerShell-Version: 7.1.0

**Comment by Windos** [comment, marked as answer]
My guess here is that the instance is spinning up, running the script, and then exiting (taking the event registration with it.)

This would be the same as if you ran the script directly, then closed the console before activating the toast notification.

Protocol handlers would work, because they are independent of the PowerShell session once they've been sent to the OS. That is, the toast triggers the handler which is registered with the OS and the OS knows how to handle it. With the events system, the toast is triggering an event INSIDE the PS session.

I'm not sure off hand, not in a position to test, but it might be possible to hold that new instance open and pop an "exit" into the action to close out the instance when it's triggered... you could end up with a bunch of orphaned sessions though depending on how you work this.

**Reply by iTob191** [reply]
> [...] it might be possible to hold that new instance open and pop an "exit" into the action to close out the instance when it's triggered... you could end up with a bunch of orphaned sessions though depending on how you work this.

Thanks for the idea! My current working approach is:
```ps1
New-BurntToastNotification `
	-Text "text" `
	-ActivatedAction {
		start "cmd" "/c echo Hello && pause"
        New-Event -SourceIdentifier "NotificationRemoved"
	}
Wait-Event -SourceIdentifier "NotificationRemoved" | Out-Null
```
This can indeed leak powershell sessions when the notification is dismissed instead of activated, but in my use-case that's not really a problem because the script is an update check and runs only once per day.
I tried also firing the NotificationRemoved event in the -DismissedAction handler to resolve this issue, but unfortunately that handler gets called in two different situations (first when you dismiss the on-screen popup, then again when you dismiss the notification from the notification center) and I found no way to distinguish those.
Anyway, the solution is sufficient for my use-case. Thanks a lot!

**Reply by Windos** [reply]
[This](https://github.com/Windos/Talks/blob/main/2020/10%20-%20October/14%20-%20SoCalPSUG%20-%20Croque%20Monsieur%20-%20Step%20up%20Your%20Toast%20Notification%20Game/Demos/8_Events.ps1) may be of interest, and [here's a video](https://www.youtube.com/watch?v=40ztRoxaS6g) of that presentation for some more context.

In short, the dismissed event has information about _why_ it was dismissed:

```powershell
$Dismissed = {
    if ($Event.SourceArgs[1].Reason -eq 'TimedOut') {
        Write-Warning 'Hey, wake up!'
    } elseif ($Event.SourceArgs[1].Reason -eq 'UserCanceled') {
        Write-Warning "I didn't want to talk to you anyway..."
    } else {
        Write-Warning "Well, this is awkward"
    }
}
```

You probably want some logic to ignore the "TimedOut" reason, as that is when it times out from the screen and goes to the Action Center. The "UserCanceled" reason is dismissing from on screen or from the action center (unfortunately I don't think you can tell from where it was dismissed).

**Reply by iTob191** [reply]
Thanks for the additional info. I think I found an easier solution though:
```ps1
-DismissedAction {
	if ($dismissedBefore) {
		New-Event -SourceIdentifier "NotificationRemoved"
	}

	$dismissedBefore = $true
}
```
This works regardless of whether the notification popup is actively dismissed by the user or by timeout. This solution is still not perfect because notifications can get removed unnoticed, either by overwriting (-> `-UniqueIdentifier`) or when the number of notifications exceeds the limit of 20 notifications per app. For my use-case the latter is impossible and the impact of the former pretty much irrelevant.

**Comment by Windos** [comment]
Hello, and thank you for your contribution to BurntToast.

Today marks the official release of [BurntToast v1.0.0](https://github.com/Windos/BurntToast/releases/tag/v1.0.0). As part of this major milestone and to ensure the project starts its next chapter with a clean slate, I am declaring "issue bankruptcy."

This decision is explained in more detail in the official announcement post: [Ten Years Toasting: BurntToast Hits the Big v1!](https://toastit.dev/2025/07/22/burnttoast-v1/)

As a result, I am closing all outstanding issues and pull requests (and discussions), including this one. Please do not interpret this as a lack of interest in the problem, suggestion, or topic you've raised. This is a necessary step to clear a very old backlog and focus development efforts on the current state of the module.

**If this topic is still relevant to you:**

1.  Please **open a new discussion**. You are welcome to link back to this one for historical context.

I hope this reset will allow me to be more responsive to community feedback going forward.

Thank you for your understanding and for being part of the BurntToast journey.
